File: test/VisualBasic/CodeCracker.Test/Design/StaticConstructorExceptionTests.vb
Imports CodeCracker.VisualBasic.Design
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Testing
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports System.IO
Imports Xunit

Namespace Design
    Public Class StaticConstructorExceptionTests
        Inherits CodeFixVerifier(Of StaticConstructorExceptionAnalyzer, StaticConstructorExceptionCodeFixProvider)

        <Fact>
        Public Async Function WarningIfExceptionIsThrownInsideStaticConstructor() As Task
            Const test = "
Public Class TestClass
    Shared Sub New()
        Throw New System.Exception()
    End Sub
End Class"

            Dim expected = New DiagnosticResult(DiagnosticId.StaticConstructorException.ToDiagnosticId(), Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(4, 9) _
                .WithMessage("Don't throw exceptions inside static constructors.")
            Await VerifyBasicDiagnosticAsync(test, expected)
        End Function

        <Fact>
        Public Async Function NotWarningWhenNoExceptionIsThrownInsideStaticConstructor() As Task
            Const test = "
Public Class TestClass
    Public Sub New()
        Throw New System.Exception()
    End Sub
End Class"

            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function StaticConstructorWithoutException() As Task
            Const test = "
Public Class TestClass
    Shared Sub New()
        
    End Sub
End Class"

            Await VerifyBasicHasNoDiagnosticsAsync(test)

        End Function

        <Fact>
        Public Async Function InstanceConstructorWithoutException() As Task
            Const test = "
Public Class TestClass
    Public Sub New()
        
    End Sub
End Class"

            Await VerifyBasicHasNoDiagnosticsAsync(test)

        End Function

        <Fact>
        Public Async Function WhenThrowIsRemovedFromStaticConstructor() As Task
            Const test = "
Public Class TestClass
    Shared Sub New()
        Throw New System.Exception()
    End Sub
End Class"

            Const fix = "
Public Class TestClass
    Shared Sub New()
    End Sub
End Class"
            Await VerifyBasicFixAsync(test, fix, 0)

        End Function

        <Fact>
        Public Async Function WhenThrowIsRemovedFromAllStaticConstructors() As Task
            Const test1 = "
Public Class TestClass1
    Shared Sub New()
        Throw New System.Exception()
    End Sub
End Class"
            Const fix1 = "
Public Class TestClass1
    Shared Sub New()
    End Sub
End Class"

            Const test2 = "
Public Class TestClass2
    Shared Sub New()
        Throw New System.Exception()
    End Sub
End Class"
            Const fix2 = "
Public Class TestClass2
    Shared Sub New()
    End Sub
End Class"

            Await VerifyBasicFixAllAsync(New String() {test1, test2}, New String() {fix1, fix2})
        End Function

        <Fact>
        Sub CanGetTypeSymbolForInferedString()
            Const code = "
    Class C
        Shared Sub Main()
            Dim b As String = """"
            Dim a = """"
        End Sub
    End Class"

            Dim tree = SyntaxFactory.ParseSyntaxTree(code)
            Dim compilation = VisualBasicCompilation.Create("test", {tree}, {MetadataReference.CreateFromFile(GetType(Object).Assembly.Location)})

            Dim result = compilation.Emit(New MemoryStream)

            Dim semanticModel = compilation.GetSemanticModel(tree)

            Dim localNodes = tree.GetRoot().DescendantNodes.OfType(Of LocalDeclarationStatementSyntax)
            For Each node In localNodes
                Dim localSym = semanticModel.GetDeclaredSymbol(node.Declarators.Single.Names.Single)
                Trace.WriteLine(localSym.ToDisplayString())

                Dim symbol = semanticModel.GetTypeInfo(node) ' Is Nothing
                Dim variableType = node.Declarators.First.AsClause?.Type ' This is null for inferred types
                If variableType IsNot Nothing Then
                    Dim typeSymbol = semanticModel.GetTypeInfo(variableType).ConvertedType
                End If
            Next
        End Sub

    End Class
End Namespace

File: test/CSharp/CodeCracker.Test/Design/StaticConstructorExceptionTests.cs
using CodeCracker.CSharp.Design;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Testing;
using System.Threading.Tasks;
using Xunit;

namespace CodeCracker.Test.CSharp.Design
{
    public class StaticConstructorExceptionTests : CodeFixVerifier<StaticConstructorExceptionAnalyzer, StaticConstructorExceptionCodeFixProvider>
    {
        [Fact]
        public async Task WarningIfExceptionIsThrowInsideStaticConstructor()
        {
            const string test = @"
                public class MyClass
                {
                    static MyClass()
                    {
                        throw new System.Exception(""error message"");
                    }
                }";

            var expected = new DiagnosticResult(DiagnosticId.StaticConstructorException.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(6, 25)
                .WithMessage("Don't throw exceptions inside static constructors.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task NotWarningWhenNoExceptionIsThrowInsideStaticConstructor()
        {
            const string test = @"
                public class MyClass
                {
                    public MyClass()
                    {
                        throw new System.Exception(""error message"");
                    }
                }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task StaticConstructorWithoutException()
        {
            const string test = @"
                public class MyClass
                {
                    static MyClass()
                    {

                    }
                }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task InstanceConstructorWithoutException()
        {
            const string test = @"
                public class MyClass
                {
                    public MyClass()
                    {

                    }
                }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task WhenThrowIsRemovedFromStaticConstructor()
        {
            const string source = @"
                public class MyClass
                {
                    static MyClass()
                    {
                        throw new System.Exception(""error message"");
                    }
                }";

            const string fixtest = @"
                public class MyClass
                {
                    static MyClass()
                    {
                    }
                }";

            await VerifyCSharpFixAsync(source, fixtest, 0);
        }

        [Fact]
        public async Task WhenThrowIsRemovedFromAllStaticConstructors()
        {
            const string source1 = @"
                public class MyClass1
                {
                    static MyClass1()
                    {
                        throw new System.Exception(""error message"");
                    }
                }
                public class MyClass3
                {
                    static MyClass3()
                    {
                        throw new System.Exception(""error message"");
                    }
                }";
            const string fixtest1 = @"
                public class MyClass1
                {
                    static MyClass1()
                    {
                    }
                }
                public class MyClass3
                {
                    static MyClass3()
                    {
                    }
                }";

            const string source2 = @"
                public class MyClass2
                {
                    static MyClass2()
                    {
                        throw new System.Exception(""error message"");
                    }
                }";
            const string fixtest2 = @"
                public class MyClass2
                {
                    static MyClass2()
                    {
                    }
                }";

            await VerifyCSharpFixAllAsync(new string[] { source1, source2 }, new string[] { fixtest1, fixtest2 });
        }
    }
}
